fem/bilininteg_ext_mass.cpp

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#include "../general/forall.hpp"
+#include "bilininteg.hpp"
+#include "gridfunc.hpp"
+
+using namespace std;
+
+namespace mfem
+{
+
+const int MAX_Q1D = 10;
+const int MAX_NBZ = 10;
+
+template<const int T_D1D = 0, const int T_Q1D = 0, const int T_NBZ = 0> static
+bool SmemPAMassApply2D(const int NE,
+                       const double* _B,
+                       const double* _Bt,
+                       const double* _op,
+                       const double* _x,
+                       double* _y,
+                       const int d1d = 0,
+                       const int q1d = 0,
+                       const int nbz = 0)
+{
+   const int D1D = T_D1D ? T_D1D : d1d;
+   const int Q1D = T_Q1D ? T_Q1D : q1d;
+   const int NBZ = T_NBZ ? T_NBZ : MAX_NBZ;
+   const int MQ1 = T_Q1D ? T_Q1D : MAX_Q1D;
+   MFEM_VERIFY(Q1D <= MAX_Q1D, "");
+   MFEM_VERIFY(NBZ <= MAX_NBZ, "");
+   const DeviceMatrix B(_B, Q1D, D1D);
+   const DeviceMatrix Bt(_Bt, D1D, Q1D);
+   const DeviceTensor<3> op(_op, Q1D, Q1D, NE);
+   const DeviceTensor<3> x(_x, D1D, D1D, NE);
+   DeviceTensor<3> y(_y, D1D, D1D, NE);
+   MFEM_FORALL_XYZ(e, NE, Q1D, Q1D, NBZ,
+   {
```

Comment:
Is this still a single lambda?

Reply:
Yes, it is. We are discussing about having different ones, but all this is still experimental.